optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/OptaPlannerRuleEventListener.java

```diff
@@ -19,25 +19,36 @@
 import org.drools.core.common.AgendaItem;
 import org.kie.api.runtime.rule.Match;
 import org.kie.internal.event.rule.RuleEventListener;
-import org.optaplanner.core.impl.score.holder.AbstractScoreHolder;
 
 public final class OptaPlannerRuleEventListener implements RuleEventListener {
 
     @Override
     public void onUpdateMatch(Match match) {
-        undoPreviousMatch((AgendaItem) match);
+        undoPreviousMatch(match);
     }
 
     @Override
     public void onDeleteMatch(Match match) {
-        undoPreviousMatch((AgendaItem) match);
+        undoPreviousMatch(match);
     }
 
-    public void undoPreviousMatch(AgendaItem agendaItem) {
-        Object callback = agendaItem.getCallback();
-        // Some rules don't have a callback because their RHS doesn't do addConstraintMatch()
-        if (callback instanceof AbstractScoreHolder.ConstraintActivationUnMatchListener) {
-            ((AbstractScoreHolder.ConstraintActivationUnMatchListener) callback).run();
+    public void undoPreviousMatch(Match match) {
+        AgendaItem<?> agendaItem = (AgendaItem) match;
+        Runnable callback = agendaItem.getCallback();
+        /*
+         * In DRL, it is possible that RHS would not call addConstraintMatch() and do some insertLogical() instead,
+         * and therefore the callback would be null.
+         * In CS-D, that is not possible, as all the rules we generate will create a match.
+         * But null callbacks can still happen.
+         *
+         * If we insert a fact and then immediately delete it without firing any rules inbetween,
+         * a dummy match will be created by Drools and that match will not have our callback in it.
+         * Although this is inefficient, it was decided that the cure would have been worse than the disease.
```

Comment:
FYI this was the cause of the downstream problem in Walter's Kogito task assigning.
(Through a series of variable listeners, something was updated and then immediately retracted, without ever firing the rules.)